### ISSUE TITLE
Develop classes Neuron, Synapse, NeuralNetwork - Feedforward code

### ISSUE THREAD

**Issue by ajovanov95** [ajovanov95/probabilistic-spiking-neural-networks#5]
Have a general relationship that Neuron and Synapse are related while both are contained in the NeuralNetwork class.

In addition construct a few factory functions to generate typical neural networks and to provide usual weight initializations (e.g. glorot, std normal, uniform etc)

More details on the implementations coming soon.

**Comment by ajovanov95** [comment]
# Planning 1

Each Neuron will have the following attributes
- `visibility` as one of `{OBSERVED, HIDDEN}`
- `bias` for the bias they have for the membrane potential function
- `id` which is the sequential integer id starting from 0

Each synapse will have the following attributes
- `neuron_from_id`
- `neuron_to_id`
- `weight` which is the strength of this synapse for the post-synaptic neuron membrane potential

The neural network would then be a graph whose nodes are `Neuron`s and whose edges are `Synapses`. The simplest way to define this graph would be the set of all neurons and set of all synapses. But this representation is neither efficient nor easy to work with for both feedforward and backpropagation directions.

---

The best approach (at least for feedforward) seems to be a use of adjacency lists for storing the connections. Each entry in this adjacency list will be the pair `neuron_to_id` and `weight`.
Thus there will be no need for a class called `Synapse`.  The `Neuron` class can be implemented as a PODO struct and a list of `Neuron`s is to be kept along the connections adjacency list.

Given those two there should be enough information to implement the feedforward process. In addition another adjacency list showing the inverted connections can come useful for the backward process. The need for the reverse adjacency list should be analyzed in more detail.

**Comment by ajovanov95** [comment]
The plan did not really turn out to be implemented as is but in general the primary lines were followed.

Due to inability to squeeze everything into one commit this issue is developed in a new branch `neural_network` starting from commit 46ca2a0

The basic functionality of constructing a neural network using functions is implemented and weakly tested using manual tests. Automatic tests should be added as next step with documentation as well along with any necessary changes in code incurred by either or by further planning.

Thus a short task list:
- [ ] Automatic unit tests for the neural network construction.
- [ ] Documentation for the neural network construction functions.
- [ ] Input checks for the neural network (e.g. weights must not be NaN, +- inf, n_input neurons must be greater than 1 etc)
- [ ] Feedforward code
- [ ] Feedforward code tests
- [ ] Feedforward code documentation
 
After these a merge with master is to happen. Following the merge a new issue is to be opened about development of the SGD gradient descent code with no-hidden layers. Following it in the longer future the training with hidden neurons is to be developed.

**Comment by ajovanov95** [comment]
The progress is steady. Before work can start on the feedforward code it must be examined in profiler why for a network of size 20 30 10 it takes 2 seconds and for a network of size 200 300 100 it takes 60 seconds to construct. This process is too slow and it must be improved without sacrificing the functional API (ideally).

https://kcachegrind.github.io/html/Home.html

That looks like a nice tool for this task.

Tasks:
- [ ] Use callgrind to see why the tests are so slow. First check if this is a Debug only problem or also happens in Release mode binaries.
- [ ]  Feedforward code
- [ ] Feedforward code tests
- [ ] Feedforward code documentation

**Comment by ajovanov95** [comment]
Feedforward is mostly done. 

Some improvements could be done to tests and possibly two more additional checks:
- The input signals contain no non-bool values. This should be checked by compiler but still. C++ is not perfectly type-safe.
- Two checks for NaNs: One in weighted-sum and one before calculating membrane potential. This could potentially slow down the code so they should be considered maybe as only running when DEBUG is enabled and not always.

I will leave these for later as there is the more important task of developing a dataset class, saving and loading of weights and the network. Finally there are the two training algorithms.

This issue will be closed for the time being as the primary goals were achieved.
